src/client/inputhandler.h

```diff
@@ -393,8 +393,42 @@ class RandomInputHandler : public InputHandler
 	virtual bool wasKeyPressed(GameKeyType k) { return false; }
 	virtual bool wasKeyReleased(GameKeyType k) { return false; }
 	virtual bool cancelPressed() { return false; }
-	virtual float getMovementSpeed() {return 0.0f;}
-	virtual float getMovementDirection() {return 0.0f;}
+	virtual float getMovementSpeed() {
+		bool f = keydown[keycache.key[KeyType::FORWARD]],
+			b = keydown[keycache.key[KeyType::BACKWARD]],
+			l = keydown[keycache.key[KeyType::LEFT]],
+			r = keydown[keycache.key[KeyType::RIGHT]];
+		if (f || b || l || r)
+		{
+			// if contradictory keys pressed, stay still
+			if (f && b && l && r)
+				return 0.0f;
+			else if (f && b && !l && !r)
+				return 0.0f;
+			else if (!f && !b && l && r)
+				return 0.0f;
+			return 1.0f; // If there is a keyboard event, assume maximum speed
+		}
+		return 0.0f;
+	}
+	virtual float getMovementDirection() {
+		float x = 0, z = 0;
+
+		/* Check keyboard for input */
+		if (keydown[keycache.key[KeyType::FORWARD]])
+			z += 1;
+		if (keydown[keycache.key[KeyType::BACKWARD]])
+			z -= 1;
+		if (keydown[keycache.key[KeyType::RIGHT]])
+			x += 1;
+		if (keydown[keycache.key[KeyType::LEFT]])
+			x -= 1;
+
+		if (x != 0 || z != 0) /* If there is a keyboard event, it takes priority */
+			return atan2(x, z);
```

Comment:
I would have expected a dedicated implementation in `RandomInputHandler::step` that chooses a random direction to go in (or to stay still) rather than just duplicating this code...

Reply:
random joystick-movement implemented. Even switching regularly between keyboard input and joystick movement